### PR TITLE
[FLINK-37925][table] Support casting from VARIANT to primitive types

### PR DESCRIPTION
**Pull request by raminqaf**
## What is the purpose of the change
                                                         
Add `CAST` and `TRY_CAST` from `VARIANT` to primitive scalar types (FLIP-521, sub-task FLINK-37925). Previously a `VARIANT` could only be read as another variant or rendered to a JSON string, with no way to extract a typed scalar out of it.
                                                                                                          
## Brief change log
                                                                                               
  - Add `VariantToPrimitiveCastRule` (registered in `CastRuleProvider`) covering BOOLEAN, all integer/floating/decimal numerics,   
BINARY/VARBINARY, DATE, TIMESTAMP and TIMESTAMP_LTZ.                                                                               
  - Numeric targets are lenient: any numeric variant converts following the regular numeric cast semantics. Non-numeric targets require the stored value to match the target kind.                                                                                 
  - On a type mismatch `CAST` fails the job and `TRY_CAST` returns `NULL`.                                                         
  - Permit the cast at planner validation time via a `VARIANT` source branch in `LogicalTypeCasts` (explicit only).                
  - Out of scope: CHARACTER_STRING (already served by the JSON display rule and `JSON_STRING`), `TIME` (no counterpart in the variant type model), and constructed types ARRAY/ROW/STRUCTURED (tracked in FLINK-37926).                                          
                                                                                                                                   
## Verifying this change                                                                                                           
                                                                                                                                   
This change added tests and can be verified as follows:                                                                            
                                                                                                                                   
- `CastRulesTest`: variant → each supported scalar, covering success, type mismatch (`CAST` throws), and null inputs.            
- `CastRuleProviderTest`: the rule resolves for supported targets.                                                               
- `LogicalTypeCastsTest`: explicit-cast support matrix, including that TIME/STRING/ARRAY and all implicit casts are rejected.    
- `CastFunctionITCase`: end-to-end `CAST`/`TRY_CAST` over `PARSE_JSON(...)`.                                                     
                                                                                                                                   
## Does this pull request potentially affect one of the following parts:                                                           
                                                                                                                                   
- Dependencies (does it add or upgrade a dependency): no                                                                         
- The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no                                              
- The serializers: no                                                                                                            
- The runtime per-record code paths (performance sensitive): no                                                                  
- Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no   
- The S3 file system connector: no                                                                                               
                                                                                                                                   
## Documentation                                                                                                                   
                                                                                                                                   
- Does this pull request introduce a new feature? yes                                                                            
- If yes, how is the feature documented? docs (VARIANT casting section in `data-types.md`) and JavaDocs                          
                                                                                                                                   
---                                                                                                                                
                                                                                                                                   
##### Was generative AI tooling used to co-author this PR?                                                                         
                                                                                                                                   
- [X] Yes

Opus 4.8

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeCasts.java [hunk 1]
```diff
@@ -726,6 +730,31 @@ private static boolean supportsConstructedCasting(
...
+            case DOUBLE:
+            case DECIMAL:
+            case BINARY:
+            case VARBINARY:
+            case DATE:
+            case TIMESTAMP_WITHOUT_TIME_ZONE:
+            case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+                return true;
+            default:
+                // TIME has no counterpart in the Variant type model. CHARACTER_STRING is handled by
```
Comment (davidradl) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeCasts.java:751]: I wonder if the cast is to a string then we could issue a helpful exception message indicating which SQL functions the user should use.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/VariantToPrimitiveCastRule.java [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    protected String generateCodeBlockInternal(
+            CodeGeneratorCastRule.Context context,
+            String inputTerm,
+            String returnVariable,
+            LogicalType inputLogicalType,
+            LogicalType targetLogicalType) {
+        final CodeWriter writer = new CastRuleUtils.CodeWriter();
+        switch (targetLogicalType.getTypeRoot()) {
+            case BOOLEAN:
+                writer.assignStmt(returnVariable, methodCall(inputTerm, "getBoolean"));
```
Comment (davidradl) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/VariantToPrimitiveCastRule.java:125]: Do variants support nullable and non nullable concepts - should a cast of of boolean when the value is null work?

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/LogicalTypeCastsTest.java [hunk 3]
```diff
@@ -262,7 +263,29 @@ private static Stream<Arguments> testData() {
...
+
+                // variant to scalar is explicit only
+                Arguments.of(new VariantType(), new BooleanType(), false, true),
+                Arguments.of(new VariantType(), new TinyIntType(), false, true),
+                Arguments.of(new VariantType(), new IntType(), false, true),
+                Arguments.of(new VariantType(), new BigIntType(), false, true),
+                Arguments.of(new VariantType(), new DoubleType(), false, true),
+                Arguments.of(new VariantType(), new DecimalType(10, 2), false, true),
+                Arguments.of(new VariantType(), new DateType(), false, true),
+                Arguments.of(new VariantType(), new TimestampType(), false, true),
```
Comment (snuyanzin) [inline flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/LogicalTypeCastsTest.java:278]: how about different precision for timestamp?

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/LogicalTypeCastsTest.java [hunk 4]
```diff
@@ -262,7 +263,29 @@ private static Stream<Arguments> testData() {
...
+                Arguments.of(new VariantType(), new LocalZonedTimestampType(), false, true),
+                Arguments.of(
```
Comment (snuyanzin) [inline flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/LogicalTypeCastsTest.java:282]: is interval intentionally skipped?

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/LogicalTypeCastsTest.java [hunk 5]
```diff
@@ -262,7 +263,29 @@ private static Stream<Arguments> testData() {
...
+                        new VariantType(),
+                        new VarBinaryType(VarBinaryType.MAX_LENGTH),
+                        false,
+                        true),
+                // variant identity cast is implicit
+                Arguments.of(new VariantType(), new VariantType(), true, true),
+                // TIME, character strings and constructed targets are not castable from variant
+                Arguments.of(new VariantType(), new TimeType(), false, false),
+                Arguments.of(new VariantType(), VarCharType.STRING_TYPE, false, false),
+                Arguments.of(new VariantType(), new ArrayType(new IntType()), false, false));
```
Comment (snuyanzin) [inline flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/LogicalTypeCastsTest.java]: map, row?

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/LogicalTypeCastsTest.java (hunk 4 above)
Comment (raminqaf) [reply flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/LogicalTypeCastsTest.java:282]: Variant.Type has no interval kind. Please refer to: https://cwiki.apache.org/confluence/spaces/FLINK/pages/349637099/FLIP-521+Integrating+Variant+Type+into+Flink+Enabling+Efficient+Semi-Structured+Data+Processing#FLIP521:IntegratingVariantTypeintoFlink:EnablingEfficientSemiStructuredDataProcessing-Casting

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/LogicalTypeCastsTest.java (hunk 5 above)
Comment (raminqaf) [reply flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/LogicalTypeCastsTest.java]: Added

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/LogicalTypeCastsTest.java (hunk 3 above)
Comment (raminqaf) [reply flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/LogicalTypeCastsTest.java:278]: Added

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/VariantToPrimitiveCastRule.java (hunk 2 above)
Comment (raminqaf) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/VariantToPrimitiveCastRule.java:125]: As far as I can see in the implementation of VARIANT type the get methods (here getBoolean) return a `boolean` and not an nullable `Boolean`.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java [hunk 6]
```diff
@@ -132,9 +133,42 @@ Stream<TestSetSpec> getTestSetSpecs() {
...
+                                call("PARSE_JSON", "42").cast(INT()),
+                                "CAST(PARSE_JSON('42') AS INT)",
+                                42,
+                                INT().notNull())
+                        .testResult(
+                                call("PARSE_JSON", "42").cast(BIGINT()),
+                                "CAST(PARSE_JSON('42') AS BIGINT)",
+                                42L,
+                                BIGINT().notNull())
+                        .testResult(
```
Comment (snuyanzin) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java:205]: dummy question: what will happen with `Double.NaN`, `Double.NEGATIVE_INFINITY`, `Double.POSITIVE_INFINITY` while transforming to VARIANT and vice versa?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/VariantToPrimitiveCastRule.java (hunk 2 above)
Comment (raminqaf) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/VariantToPrimitiveCastRule.java:125]: Whereas `NULL` type is supported in VARIANT type. I have added logic for this and more tests to cover NULL casts and nullability.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeCasts.java (hunk 1 above)
Comment (raminqaf) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeCasts.java:751]: Even though this is a good idea and helpful for the user. Implementing it is a bit messy. Still I kept the changes for you to review. Since it adds lots of nose in the method. But let me know what you think. See `getUnsupportedCastHint` and `CastInputTypeStrategy`.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java [hunk 7]
```diff
@@ -132,9 +133,65 @@ Stream<TestSetSpec> getTestSetSpecs() {
...
+        return Collections.singletonList(
+                TestSetSpec.forExpression("Cast a VARIANT produced by PARSE_JSON to a primitive")
+                        .onFieldsWithData("unused")
+                        .andDataTypes(STRING())
+                        .testResult(
+                                call("PARSE_JSON", "42").cast(INT()),
+                                "CAST(PARSE_JSON('42') AS INT)",
+                                42,
+                                INT().notNull())
+                        .testResult(
```
Comment (snuyanzin) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java:200]: also probably would make sense to have a couple of tests with overflow

Comment (snuyanzin) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java:200]: looks like table api and and SQL process overflow different, I would say it is a bug...

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java [hunk 8]
```diff
@@ -132,9 +133,65 @@ Stream<TestSetSpec> getTestSetSpecs() {
...
+        specs.addAll(variantCasts());
         return specs.stream();
     }
 
+    private static List<TestSetSpec> variantCasts() {
+        // A variant is produced with PARSE_JSON so that the source column stays a STRING literal;
+        // there is no VARIANT literal to feed a source column directly. A JSON integer is stored in
+        // the smallest integer type that fits (42 -> TINYINT), and numeric casts are lenient, so it
+        // still widens to INT.
+        return Collections.singletonList(
```
Comment (snuyanzin) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java]:
```suggestion
        return List.of(
```

Comment (raminqaf) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java]: Addressed

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java (hunk 6 above)
Comment (raminqaf) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java:205]: I got an exception on the `Double.NEGATIVE_INFINITY`
```
Caused by: org.apache.flink.shaded.jackson2.com.fasterxml.jackson.core.JsonParseException: Non-standard token '-Infinity': enable `JsonReadFeature.ALLOW_NON_NUMERIC_NUMBERS` to allow
 at [...]
```

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java (hunk 7 above)
Comment (raminqaf) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java:200]: @snuyanzin Could you please elaborate how you found the bug? I couldn't really reproduce a problematic output here. I have added two tests that verifies that the overflow work as expected.

Comment (snuyanzin) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java:200]: just add a test like
```java
.testResult(
                                call("PARSE_JSON", "422").cast(TINYINT()),
                                "CAST(PARSE_JSON('422') AS TINYINT)",
                                422,
                                TINYINT().notNull())
```

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java (hunk 6 above)
Comment (snuyanzin) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java:205]: I like it in every places it fails differently 🙈

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java (hunk 7 above)
Comment (raminqaf) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java:200]: The expected `422` is wrong. `TINYINT` is a signed byte (−128..127). It physically cannot hold 422. The cast wraps: `(byte) 422 = -90`. The correct case is:                                                                                                          
 ```java                                                                                                                                  
  .testResult(                                                                                                                       
          call("PARSE_JSON", "422").cast(TINYINT()),                                                                                 
          "CAST(PARSE_JSON('422') AS TINYINT)",                                                                                      
          (byte) -90,                                                                                                                
          TINYINT().notNull())
```

Comment (raminqaf) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java:200]: @snuyanzin I have added more tests for different types and assess their overflow behavior.

**Review by snuyanzin — APPROVED** [review]
